### ISSUE TITLE
Trilinos:  Avoid Circular Link Dependencies

### ISSUE THREAD

**Issue by jjellio** [trilinos/Trilinos#5133]
## Enhancement
@trilinos

Several issues keep cropping up that are related to circular dependencies in libraries or poor link order (e.g., libB.a depends on libA.a, but is linked as libA.a libB.a

A standard approach for handing these issues is to wrap libraries inside a 'linker group'.

E.g., See Intel's Link Adviser (this is how MKL handles circular dependencies in their lib)

To handle this, we need CMake support, as well as some fixes in Trilinos' library finding algorithms.

What are linker groups and how do they work?  Effectively, a linker group starts with --start-group ..list your libraries here ... --end-group.  When the linker encounters a symbol it needs to locate, it scans from the start of the link line towards the end. (left to right).  When it encounters a linker group, it will repeatedly scan the group until it determines the symbol is clearly not defined within the group. Then it moves right. (to the next lib, object file, or linker group).

Downside: Linker groups were initially given grief, because they can cause the linker to spend lots of time scanning when the symbol is clearly in the subsequent library.

Trilinos' upside:  From my experience, hacking these groups in results in much smaller build files, and it doesn't really slow the linker down: Why? Because CMake automatically adds libraries with a 'multiplicity' of 2, and CMake explicitly creates a link line with the appropriate libraries dependencies listed always.  This means, if Tpetra depends on libaries A,B,C, and MueLu Depends on Tpetra, A, B, C, you will get a link line that looks like: -lTpetra -lA -lB -lC -lMueLu -lA -lB -lC.  With linker groups, we could do: --start-group <Trilinos Libs Here> --end-group --start-group ...TPL_A.. --endgroup --start-group TPL_B ... --end-group.

The above is contrived, but with actual Trilinos based apps this results in a *massive* reduction in replicated libraries on the link line, which also means your CMake created build files (build.ninja or link.txt) are much much smaller.

Trilinos Upside 2:  Circular issues.  With groups, the fact libraries have circular references goes away at the cost of potentially longer link times... Read on:  One way to evaluate the perceived penalty is to hack the link line for an app, and see what practical slowdown groups make... from my experience it was none, and you gain protection against circular library dependencies.

How would this work?
  1.  TPLs get automatically wrapped with a linker group
    a.  Why? This avoids having to replicate the TPL libraries multiple times
  2.  Trilinos's libraries get wrapped inside a linker group
    a.  You can hack-this effect by modifying the lib/cmake/Trilinos.cmake installed cmake file.
    b.  The benefit is mostly for *installed* Trilinos, That is, Trilinos that is actually used by someone. (not developer R & D)

What breaks this:
  1. You can't specify TPL libraries in a linker group, e.g., BLAS_LIBRARIES="-Wl,--start-group;lib1;lib2;lib3;-Wl,--end-group"
    a.  Trilinos' logic currently interprets commas as semicolons (perhaps this is in Tribits?) - which ends up with CMake breaking the linker option. Raw CMake will accept linker options in the library list, and it ignores/leaves in place anything that starts with a '-', which allows you provide linker options in your library definition.

I'd like to start a dialog on using this.  This is Linux/GNU Linker style thing, but most of our customers use Trilinos on machines that build using regular GNU type tools (LD, Gold). LLVM's LLD advertises that they do this natively (they claim to record symbols as they see them, and automatically link symbols they have seen (so they don't need linker groups at all).

@bartlettroscoe

**Comment by bartlettroscoe** [comment]
See:

* https://en.wikipedia.org/wiki/Acyclic_dependencies_principle
* https://www.youtube.com/watch?v=gQx4HV9s_h0

Note that languages like Python and Java don't allow circular depends between modules/packages.

**Comment by mhoemmen** [comment]
@jjellio Just curious -- what's a concrete example of Trilinos having circular link dependencies?

**Comment by bartlettroscoe** [comment]
In more detail ...

Acyclic dependencies is one of the most fundamental software engineering packaging principles.  For example, see the “Acyclic dependencies principle”:

*    https://en.wikipedia.org/wiki/Acyclic_dependencies_principle
*    https://khalilstemmler.com/wiki/acyclic-dependencies-principle/
*    https://link.springer.com/chapter/10.1007/978-1-4842-4119-6_9
   
The reference:

*   https://khalilstemmler.com/wiki/acyclic-dependencies-principle/

summarizes this well:

* Make the code hard to understand, thus reducing maintainability
* Reduce the quality of the code and improve the likelihood of errors by making it harder to change one component without affecting another
* Make it impossible to test each individual component because they can't be separated
* Have the ability to violate [The Dependency Rule](https://khalilstemmler.com/wiki/dependency-rule/)

Also, from a practical standpoint, you can’t have circular dependencies between shared libraries and CMake does not technically support circular dependencies between shared libraries, only static libraries (see here).    Therefore, I think Trilinos would have to drop support for shared libraries if we want the code to be portable with circular library dependencies.

And even if you stuck with static libraries what can support circular dependencies, it would be a nightmare to implement in TriBITS.

Some other things to consider:

* It is not clear if C++20 modules are going to support circular dependencies but [this presentation](https://indico.cern.ch/event/682267/contributions/2795954/attachments/1569847/2475832/Runtime_C_Modules_presentation.pdf) suggests not.  (And even if it is possible, [this post](https://stackoverflow.com/questions/60892272/forward-declaration-to-break-cyclic-dependency-in-c20-modules-doesnt-work) suggests that it will be very difficult to support.)
* Spack does not support circular dependencies between Spack packages.  Some people have stated a desired goal to allow individual Trilinos packages to be built and installed as separate Spack packages.  (So all packages in the circular dependency would have to be considered a single Spack package.)
* Many other languages with a packaging/module system don’t even support circular dependencies.  For example, the Go language does not allow circular imports.

And you can break every circular dependency by using one of the following approaches (see [here](https://en.wikipedia.org/wiki/Acyclic_dependencies_principle#Cycle_breaking_strategies)):

* Refactor the code using [dependency inversion](https://en.wikipedia.org/wiki/Dependency_inversion_principle) and [dependency injection](https://stackify.com/dependency-injection/), or
* Move the troublesome code into a new package (e.g. the [xSDKTrilinos](https://github.com/trilinos/xSDKTrilinos) packages breaks a circular dependency between Trilinos and PETSc), or
* Merge all of the packages in the cycle in to one big package (only practical for small cycles due to link size limits)

We have used dependency inversion and dependency injection to break several cycles in Trilinos over the years (e.g. TeuchosCore and TeuchosComm).

All of these problems are avoided by just making Trilinos one big library (except it would be too big to link to on many systems).

**Comment by jjellio** [comment]
Ross, When I posted this, it isn't that there is necessarily a circular dependency (there *is* if you link MKL by hand).  It's that you have a link line where you can end up with duplicate library paths.  Trilinos deals with this currently by removing duplicates.  E.g., `-lteuchos_numerics -llapack -lblas -ltpetra -lkokkos-kernels -lkokkos -llapack -lblas` (where Cmake defaults to enforcing the proper ordering for every dependency).  The duplicate removal is somewhere in the Cmake logic (and app teams do this as well)

The linker groups just make it easier and safer than using 'remove_duplicates'.  `-Wl,--start-group -lteuchos_numerics -ltpetra -lkokkos-kernels -lkokkos -llapack -lblas -Wl,--end-group` or `-Wl,--start-group trilinos-libs -Wl,--end-group` + `-Wl,--start-group all_needed_tpls -Wl,--end-group` (these are pretty easy to form by taking the concatenation of all TPL library lists (and all trilinos library lists) and then wrapping them appropriately.

IMO, the group stuff is really just an easier and shorter way to express the link lines w/duplicates removed. It has a side effect of tolerating circular dependencies as well.

**Comment by jhux2** [comment]
So the TL;DR is that the link line can be significantly shortened by using link groups?  This seems like a no-brainer.

**Comment by jjellio** [comment]
Yes. IMO, it is safer than doing `remove duplicates` on a library list.  Both are making an implicit assumption that symbols will not be duplicately defined.

I did this mainly to shorten my static link lines on Mutrino, because they made my console scroll.

What groups actually does is tell the linker to loop over the libraries in the group until the number of undefined symbols is the same as when it started that pass. It could be slower, but I haven't noticed it. FYI, Intel recommends this approach for MKL.

It wouldn't be too hard to add this logic into the `Package_name.cmake` (that's where I did it).

**Comment by bartlettroscoe** [comment]
@jjellio, after the TriBITS refactoring in https://github.com/TriBITSPub/TriBITS/issues/299 this might be resolved.
